### PR TITLE
Features/clean typescript

### LINKED ISSUE #5: React / TS Feedback
- [x] 1. The way the repo is structed, I guess we want to package all our controls in one pcf control solution? I think that’s alright for now, just wanted to confirm if that it is the intent.
- [x] 2. Add a quick read.me to build/run project and PCFs
- [x] 3. To reduce the amount of fetches, is it possible to load the "tag" data from getRelatedEntities?
- [x] 4. ~For loadData can we clarify in a comment what it loads? if it loads more than essential can we streamline.~
- [x] 5. For associate or dissociate, have we tried one of them via context.webapi?
- [x] 6. With the queries it feels like the tagpickerbasecomponent is violating single responsibility principle by manual building those queries. If we can’t use webapi, maybe move those as an extension of WebAPI
- [x] 7. Localize component text properties
- [x] 8. Why is GetOuput in both index.ts. It seems it related to the proxy textbox field, can it be document through comments
- [x] 9. The two manifests are not including the same files, why is that the case?
- [x] 10. Variables section looks a bit busy with no comments in base component. It seems like all variables are required, can they be grouped and its usage commented on.

### PR DESCRIPTION
**Pull request by joshcampbell191**
Closes #5

### CODE REVIEW

File: src/Components/TagPickerComponent/TagPickerComponent/TagPickerBaseComponent.ts [hunk 1]
```diff
@@ -2,244 +2,259 @@ import * as React from 'react';
...
+        onEmptyInputFocus: this.onEmptyInputFocus.bind(this),
+        onResolveSuggestions: this.onResolveSuggestions.bind(this)
+    }
 
-    public selectedItems: ITag[];
+    /**
+     * Helper methods used to access common metadata properties.
+     */
+    private get idAttribute(): string { return this.relatedEntityMetadata ? this.relatedEntityMetadata[EntityMetadataProperties.PrimaryIdAttribute] : ""; }
+    private get nameAttribute(): string { return this.relatedEntityMetadata ? this.relatedEntityMetadata[EntityMetadataProperties.PrimaryNameAttribute] : ""; }
```
Comment (pierluc-ssoucy) [inline src/Components/TagPickerComponent/TagPickerComponent/TagPickerBaseComponent.ts:59]: forgot to mention this in the previous review, but this gives an error/warning saying the target should be es5 or greater. Modify tsconfig accordingly.

File: src/Components/TagPickerComponent/TagPickerComponent/TagPickerBaseComponent.ts [hunk 2]
```diff
@@ -2,244 +2,259 @@ import * as React from 'react';
...
+
+    /**
+     * Disassociate the item with the entity.
+     * @param item The item to disassociate.
+     */
+    private disassociateItem(item: ITag): Promise<Response> {
+        const parentSetName: string = this.entityMetadata[EntityMetadataProperties.EntitySetName];
+
+        return this.webAPI.disassociateRecord(parentSetName, this.entityId, this.relationshipName, item.key);
+    }
```
Comment (pierluc-ssoucy) [inline src/Components/TagPickerComponent/TagPickerComponent/TagPickerBaseComponent.ts]: now that the code is so simple, we can probably move that back to the for loop.

File: src/Components/TagPickerGridComponent/TagPickerGridComponent/TagPicker.tsx [hunk 3]
```diff
@@ -9,103 +8,103 @@ import { initializeIcons } from 'office-ui-fabric-react/lib/Icons';
...
-  tagDisplayName?: string,
-  selectedItems?: ITag[],
-  onChange?: (items?: ITag[]) => void;
-  onEmptyInputFocus?: (selectedItems?: ITag[]) => Promise<ITag[]>;
-	onResolveSuggestions?: (filter: string, selectedItems?: ITag[]) => Promise<ITag[]>;
+    inputLabel?:string;
+    noResultsFoundLabel?: string;
+    removeButtonLabel?: string;
+
+    labelText?: string;
```
Comment (pierluc-ssoucy) [inline src/Components/TagPickerGridComponent/TagPickerGridComponent/TagPicker.tsx]: we could move this into a sub property to end up with something like this.props.labels.noresultsfound

File: src/Components/TagPickerGridComponent/TagPickerGridComponent/TagPickerBaseComponent.ts [hunk 4]
```diff
@@ -2,247 +2,295 @@ import * as React from 'react';
...
+            translations.set(language, label);
+        }
+
+        const languageId: number = (<any>this.context).userSettings.languageId;
+        const userLanguage: Languages = (<any>Languages)[languageId];
+
+        // check if the user language is supported, otherwise default to English
+        const labelLanguage: number = userLanguage ? languageId : Languages.en;
+
+        return translations.get(labelLanguage) ?? "";
```
Comment (pierluc-ssoucy) [inline src/Components/TagPickerGridComponent/TagPickerGridComponent/TagPickerBaseComponent.ts]: this becomes slightly more confusing than intended with all the casting. If there's a way to make it clearer that would be great. Either way, at least adding expected format of the string would be helpful, perhaps also throw that in the an usage readme.

File: src/Components/TagPickerGridComponent/TagPickerGridComponent/TagPickerBaseComponent.ts [hunk 5]
```diff
@@ -2,247 +2,295 @@ import * as React from 'react';
...
+
+            // check if the language is supported
+            if (language === undefined)
+                continue;
+
+            const label: string = values[1];
+            translations.set(language, label);
+        }
+
+        const languageId: number = (<any>this.context).userSettings.languageId;
```
Comment (pierluc-ssoucy) [inline src/Components/TagPickerGridComponent/TagPickerGridComponent/TagPickerBaseComponent.ts]: i don't think the cast is necessary on this one.

**Review by pierluc-ssoucy — CHANGES_REQUESTED** [review]

File: src/Components/TagPickerGridComponent/TagPickerGridComponent/TagPickerBaseComponent.ts (hunk 5 above)
Comment (joshcampbell191) [reply src/Components/TagPickerGridComponent/TagPickerGridComponent/TagPickerBaseComponent.ts]: It has been removed

File: src/Components/TagPickerComponent/TagPickerComponent/TagPickerBaseComponent.ts (hunk 2 above)
Comment (joshcampbell191) [reply src/Components/TagPickerComponent/TagPickerComponent/TagPickerBaseComponent.ts]: The methods have been moved inside the loops

File: src/Components/TagPickerComponent/TagPickerComponent/TagPickerBaseComponent.ts (hunk 1 above)
Comment (joshcampbell191) [reply src/Components/TagPickerComponent/TagPickerComponent/TagPickerBaseComponent.ts:59]: This appears to be an issue with the tooling as the target is inherited from node_modules/pcf-scripts/tsconfig_base.json which includes a target of es5.

File: src/Components/TagPickerGridComponent/TagPickerGridComponent/TagPicker.tsx (hunk 3 above)
Comment (joshcampbell191) [reply src/Components/TagPickerGridComponent/TagPickerGridComponent/TagPicker.tsx]: The label properties have been moved into a sub-property (labels) of type ITagPickerLabelProps.

File: src/Components/TagPickerGridComponent/TagPickerGridComponent/TagPickerBaseComponent.ts (hunk 4 above)
Comment (joshcampbell191) [reply src/Components/TagPickerGridComponent/TagPickerGridComponent/TagPickerBaseComponent.ts]: Updated README accordingly

**Review by pierluc-ssoucy — APPROVED** [review]

**Pull request merged**